feat(CI/CD + Coveralls)

camilamacedo86 opened the pull request:
## Motivation 

## What
- Replace Trevis by Circle Ci
- Add automatic tag release process
- Add tag release and image process steps in the README
- Add health badges and coveralls

## Why
- Following the standard (https://github.com/aerogear/standards)
- Create the images automatically which is a requirement for the projects
- Good practices. 
- Add the first unit test files

## Verification Steps
- Check that it worked with success here: https://circleci.com/gh/camilamacedo86/unifiedpush-operator/6
- Check that the coveralls were added with success as well. 

## Checklist:

- [X] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [X] Finished task
- [ ] TODO

## Additional Notes
<img width="1161" alt="Screenshot 2019-07-02 at 16 25 03" src="https://user-images.githubusercontent.com/7708031/60540675-1d548400-9ce6-11e9-9e4b-2551724ea4ce.png">

camilamacedo86 committed:
feat(CI/CD) : replace trevis by Circle CI and release tag process, add health checks in the README as well

camilamacedo86 commented:
Hi @damienomurchu could you help me to check it?

camilamacedo86 committed:
feat(CI+Redame+Coveralls+UnitTest) : replace trevis by Circle CI, add info in readme, add unit tests and coveralls

grdryn reviewed (dismissed):
The images for this repo are built by Jenkins, and are only pushed after the end-to-end tests have been run. Neither Travis nor CircleCI nor anything else should be pushing an image. Given that, I'm not sure what the value of moving from Travis to CircleCI is?

camilamacedo86 commented:
Hi @grdryn, 

Thank you for your reply. Following the comments inline. 

> The images for this repo are built by Jenkins and are only pushed after the end-to-end tests have been run. 

I understand that images should be just be pushed in 2 situations as follows:

1. When the PR is merged in the master ( master and latest tag )
2. When a new tag is a released. ( E.g 2.1.0 ) 

**Is Jenkins working in this way?** 
**Have it done in the Jenkins will not reduce the maintenance ability?** 

PS. IMHO: The repo <https://quay.io/repository/aerogear/unifiedpush-operator?tab=tags> shows very polluted and has a master and latest but no tag release at all. We should clean it and not use for our tests since it is public.  

> Neither Travis nor CircleCI nor anything else should be pushing an image. Given that, I'm not sure what the value of moving from Travis to CircleCI is?

So far, I understand that: 

* We should use CircleCI and no  longer Trevis ( it was defined in the past )
* The biggest part of the projects are creating the tags by the CI because of this it is how it was done in the MSS and MSS Oper. 

Also, it is important to highlight that we did here as well:
* doc how to do a release here
* added the coveralls
* added the badges 
* added an example/first step to start to impl the unit tests.  

Hi @damienomurchu @craicoverflow,
 
You both checked deeper it, could you please provide your input as well? 
 
c/c @aliok @laurafitzgerald @psturc @danielpassos

aliok commented:
I think @psturc is the person that can give more info but I think it is a better approach to publish images with our private Jenkins instead of Travis/CircleCI.

Travis/CircleCI can be used to run unit tests though.

And I am +1 on using CircleCI instead of Travis.

Thanks for the documentation around how to make a release and things like that. Let's address the concerns above, update the docs accordingly and merge.

camilamacedo86 commented:
Hi @aliok,

Really tks for your input.

Following my argumentations against the usage of Jenkins to create the images:

- When we upgrade the version of the operator-sdk then we need to change the project + Jenkins.
- The code is not centralized and is not clear how the things are working on. 
- Reduces the maintenance ability
- I understand that the other new projects are using the CI to do this job then we will not follow the same standard which makes harder for us to keep them maintained either. 

Hi @damienomurchu and @psturc, 

Could you help us and give your inputs? 
Also, if you decided to move forward with Jenkins we need to ensure that it is working as described follows and I need to know if it is also creating the images based in the new tags pushed to the repo. 

- When the PR is merged in the master ( master and latest tag )
- When a new tag is pushed to the repo. ( E.g 2.1.0 )

craicoverflow commented:
> Hi @grdryn,
> 
> Thank you for your reply. Following the comments inline.
> 
> > The images for this repo are built by Jenkins and are only pushed after the end-to-end tests have been run.
> 
> I understand that images should be just be pushed in 2 situations as follows:
> 
>     1. When the PR is merged in the master ( master and latest tag )
> 
>     2. When a new tag is a released. ( E.g 2.1.0 )
> 
> 
> **Is Jenkins working in this way?**
> **Have it done in the Jenkins will not reduce the maintenance ability?**
> 
> PS. IMHO: The repo [quay.io/repository/aerogear/unifiedpush-operator?tab=tags](https://quay.io/repository/aerogear/unifiedpush-operator?tab=tags) shows very polluted and has a master and latest but no tag release at all. We should clean it and not use for our tests since it is public.
> 
> > Neither Travis nor CircleCI nor anything else should be pushing an image. Given that, I'm not sure what the value of moving from Travis to CircleCI is?
> 
> So far, I understand that:
> 
>     * We should use CircleCI and no  longer Trevis ( it was defined in the past )
> 
>     * The biggest part of the projects are creating the tags by the CI because of this it is how it was done in the MSS and MSS Oper.
> 
> 
> Also, it is important to highlight that we did here as well:
> 
>     * doc how to do a release here
> 
>     * added the coveralls
> 
>     * added the badges
> 
>     * added an example/first step to start to impl the unit tests.
> 
> 
> Hi @damienomurchu @craicoverflow,
> 
> You both checked deeper it, could you please provide your input as well?
> 
> c/c @aliok @laurafitzgerald @psturc @danielpassos

Hi @camilamacedo86 - I didn't do any CI work at all in the MSS operator so I can't give experience based advice on this. @damienomurchu did most if not all of it!

grdryn commented:
The reason there are no images with release tags is because we haven't done any releases yet. The Jenkins pipeline job will build tags also though.

In the sense that the most valuable thing here is the e2e tests, the Jenkins job is the one that really matters, and neither Travis nor CircleCI really do at all (the only reason I see for having either of them at all is that we don't expost the Jenkins build logs to a publicly accessible location yet). Given that, I still don't understand what value we get from migrating from one to the other. There are other projects here in the AeroGear org that use Travis too, and that's generally fine, in my opinion.

psturc commented:
> Hi @damienomurchu and @psturc,
> 
> Could you help us and give your inputs?
> Also, if you decided to move forward with Jenkins we need to ensure that it is working as described follows and I need to know if it is also creating the images based in the new tags pushed to the repo.
> 
> * When the PR is merged in the master ( master and latest tag )

This is already covered by jenkins, see [these lines](https://github.com/aerogear/unifiedpush-operator/blob/master/Jenkinsfile#L132-L157)
> * When a new tag is pushed to the repo. ( E.g 2.1.0 )

Jenkins is set up to pick up the tags as well, I can give it a try. **Update**: https://mobile-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/unifiedpush-operator/view/tags/job/test-tag/

Also Jenkins takes care of running e2e tests which is brings the most value here.

If we should get rid of travis job and migrate it to centralci, I'm ok with that. But maybe we could think about removing the other PR checks (except Jenkins) completely, since Jenkins job could take care of everything here.

camilamacedo86 commented:
HI @grdryn,

I understand that has many reasons for being defined that we should use CircleCi instead of Trevis. Also, if you check the main part of projects are already with the CircleCI. (Mainly just the old ones are using trevis). See https://github.com/search?l=YAML&q=org%3Aaerogear+circleci&type=Code  

We just need check with @psturc and @damienomurchu their inputs and if we keep the tags built by Jekins and then a change this point here or if they agree in move forward and allow us do it by the CI which will bring in POV many advanatges as described above. The rest of the changes shows 100% for all reviews so far.

damienomurchu commented:
As per comments by others - if all CI image and testing needs are handled by Jenkins, what is the requirement for having TravisCI or CircleCI?

In that case I don't see the point of migrating from Travis to Circle, and instead we should be looking at removing the Travis Ci integration. 

Am I missing anything?

camilamacedo86 commented:
HI @psturc

Are the Jenkins using the make commands? 
What make commands? How we be sure that we will not break them?
Could you provide it for I am able to at least add this info in the makefile for we do not have them? 
Can you provide the repo with the code impl of the Jenkins for we doc it as well?

Regards:

> Also Jenkins takes care of running e2e tests which is brings the most value here.

* We can add it here as well 
* How can anyone check it's logs? 
* How can we ensure that we will not remove or change something?
* How to allow anyone upgrades the operator-sdk version in the project +jenkins? 

> If we should get rid of travis job and migrate it to centralci, I'm ok with that. But maybe we could think about removing the other PR checks (except Jenkins) completely, since Jenkins job could take care of anything here.

* How can anyone check it's logs? 
* How we any dev improve/maintained the coveralls and related make commands? 
* How can we ensure that we will not remove or change something?
* How to allow anyone upgrades the operator-sdk version in the project +jenkins?

camilamacedo86 commented:
Hi @damienomurchu good point. 

I am against to move forward with the Jenkins solution because:

1 - We can remove/change make commands which are used by it since it not clear at all
2 - We cannot check the logs and errors
3 - It is hard to keep maintained since the code is not centralized. 
4 - It is definitely is a blocker to allow devs from the community 

Questions: 

1 - How for example, upgrade the operator-sdk version or change the coveralls if it is in the Jenkins?
2 - How to know what is impl in the Jenkins and in the CI here, since each part is one place and we have no visibility at all? 
3 - How to check the logs? What makes commands are used by it? How to maintain? Where is the code? 
4 - Why the trevis was added if you would like to use the Jenkins, instead of? 

NOTE: I cannot check the Jenkins and this logs at all. As a dev, I'd like to do it.

* if we move with Jenkins we need to know what is used by it, where is the code to do changes there in order indeed we are able to maintain it, and doc all. 

c/c @psturc @grdryn @aliok

damienomurchu commented:
@camilamacedo86 is there a compelling reason not to just use Travis/ CircleCI just to run the test suite for PRs and leave Jenkins in place for everything else as suggested above?

psturc commented:
@camilamacedo86 
> NOTE: I cannot check the Jenkins and this logs at all. As a dev, I'd like to do it.

You should have access to the jenkins instance if you're on VPN, I checked yesterday. Try kerberos.

> * if we move with Jenkins we need to know what is used by it, where is the code to do changes there in order indeed we are able to maintain it, and doc all.

That Jenkins instance we're using is internal and it won't give a detailed feedback to developer outside of Red Hat about the build, that's true. But since we're actively working on it atm, we can collaborate with developers outside RH which run into problems.

@darahayes mentioned the option of using ["Orbs" in circleci](https://circleci.com/orbs/registry/orb/circleci/redhat-openshift), which could help us with running e2e tests. If it proves to be stable, then we could potentially move the stuff out of Jenkins in a future.

But I don't see a big priority in moving CI/CD stuff to circleci **now**.

grdryn commented:
> @camilamacedo86
> 
> > NOTE: I cannot check the Jenkins and this logs at all. As a dev, I'd like to do it.
> 
> You should have access to the jenkins instance if you're on VPN, I checked yesterday. Try kerberos.
> 
> > * if we move with Jenkins we need to know what is used by it, where is the code to do changes there in order indeed we are able to maintain it, and doc all.
> 
> That Jenkins instance we're using is internal and it won't give a detailed feedback to developer outside of Red Hat about the build, that's true. But since we're actively working on it atm, we can collaborate with developers outside RH which run into problems.

(Yes, and we can eventually have something that uploads the logs to a public location.)

Also as an additional answer to the question in this bullet point, the code for the job is in this repo in the file `Jenkinsfile`.

> 
> @darahayes mentioned the option of using ["Orbs" in circleci](https://circleci.com/orbs/registry/orb/circleci/redhat-openshift), which could help us with running e2e tests. If it proves to be stable, then we could potentially move the stuff out of Jenkins in a future.

We _could_, but we'd still be losing something: we can trust artifacts built by our own systems more than we can trust artifacts built by <free-3rd-party-service-run-by-vc-funded-startup>. The importance of that is different depending on who you ask. I can provide examples of where things very like that has bitten projects in the past if you want, but that's a different discussion.

camilamacedo86 commented:
Hi @grdryn, @psturc, @aliok, @damienomurchu 

Really thank you for all your explanations so far and inputs. So, let's keep what is done by the Jenkins there and I'd ask for we move forward here in order to:

- Add the doc in the README over CI/CD (Check its changes)
- Add the doc in the README how to release new tags and build of images  (Check its changes)
- Start to dod the Architecture section into the README (Check its changes)
- Make clear what make command is used in each part (Check its Makefile)
- Add coveralls (Check this PR)
- Add unit tests Check this PR)
- Add badges  in the README (Check its changes)
- Replace the Trevis by The Circle CI and move forward following the standards and definitions made so far and to keep it as the other projects. 

<img width="1000" alt="Screenshot 2019-07-03 at 09 28 41" src="https://user-images.githubusercontent.com/7708031/60591504-fb561280-9d74-11e9-92c6-2d3806c395fd.png">

Wdyt? Could we move forward in this way?

grdryn commented on README.adoc:
```diff
@@ -461,6 +471,65 @@ operator-sdk build $IMAGE --enable-tests && docker push $IMAGE
+. Create a new version tag following the http://semver.org/spec/v2.0.0.html[semver], for example `0.1.0`
+. Bump the version in the link:./version/version.go[version.go] file.
+. Update the the link:./CHANGELOG.MD[CHANGELOG.MD] with the new release.
+. Looking for the SOPs and update the tag for the them in all files (e.g `https://github.com/aerogear/unifiedpush-operator/blob/0.1.0/SOP/SOP-operator.adoc`)
```
```suggestion
. Update any tag references in all SOP files (e.g `https://github.com/aerogear/unifiedpush-operator/blob/0.1.0/SOP/SOP-operator.adoc`)
```

camilamacedo86 replied:
👍

grdryn commented on README.adoc:
```diff
@@ -461,6 +471,65 @@ operator-sdk build $IMAGE --enable-tests && docker push $IMAGE
++
+NOTE: The image with the tag will be created and pushed to the https://quay.io/repository/aerogear/unifiedpush-operator[unifiedpush-operator image hosting repository] by the Jenkins.
+++
+WARNING: Do not use letters in the tag such as `v`. It will not work.
```
Why would it not work? I'd have expected any tags to work :thinking:

_If_ it's true that it wouldn't work, then I think this warning would be better placed immediately above where the instructions to create the tag are described, rather than 15 lines later.

camilamacedo86 replied:
It does not work in the previous impl. I will remove from here because I do not know in Jenkins.

grdryn commented on README.adoc:
```diff
@@ -461,6 +471,65 @@ operator-sdk build $IMAGE --enable-tests && docker push $IMAGE
+
+==== CircleCI
+
+Following the what is performed by it.
```
```suggestion

```

camilamacedo86 replied:
👍

grdryn commented on README.adoc:
```diff
@@ -461,6 +471,65 @@ operator-sdk build $IMAGE --enable-tests && docker push $IMAGE
+
+==== Jenkins
+
+Following the what is performed by it.
```
```suggestion

```

camilamacedo86 replied:
👍

camilamacedo86 commented:
@grdryn are all done. Could we move forward now?

camilamacedo86 commented:
Hi @grdryn really tks for your help and support here. 
As the help and input of all others :-) c/c @aliok @damienomurchu @psturc
